src/mixer/basetrackplayer.cpp

```diff
@@ -139,32 +141,40 @@ BaseTrackPlayerImpl::BaseTrackPlayerImpl(
 
     // Waveform controls
-    // This acts somewhat like a ControlPotmeter, but the normal _up/_down methods
-    // do not work properly with this CO.
-    m_pWaveformZoom =
-            std::make_unique<ControlObject>(ConfigKey(getGroup(), "waveform_zoom"));
-    m_pWaveformZoom->connectValueChangeRequest(this,
-            &BaseTrackPlayerImpl::slotWaveformZoomValueChangeRequest,
-            Qt::DirectConnection);
-    m_pWaveformZoom->set(1.0);
-    m_pWaveformZoomUp = std::make_unique<ControlPushButton>(
-            ConfigKey(getGroup(), "waveform_zoom_up"));
-    connect(m_pWaveformZoomUp.get(),
-            &ControlPushButton::valueChanged,
-            this,
-            &BaseTrackPlayerImpl::slotWaveformZoomUp);
-    m_pWaveformZoomDown = std::make_unique<ControlPushButton>(
-            ConfigKey(getGroup(), "waveform_zoom_down"));
-    connect(m_pWaveformZoomDown.get(),
-            &ControlPushButton::valueChanged,
-            this,
-            &BaseTrackPlayerImpl::slotWaveformZoomDown);
-    m_pWaveformZoomSetDefault = std::make_unique<ControlPushButton>(
-            ConfigKey(getGroup(), "waveform_zoom_set_default"));
-    connect(m_pWaveformZoomSetDefault.get(),
-            &ControlPushButton::valueChanged,
+    // Unfortunately, since WaveformWidgetFactory isn't instantiated yet, we
+    // can't use WaveformWidgetFactory::instance()->getDefaultZoom()
+    // Read from config ourselves instead.
+    double defaultZoom = m_pConfig->getValue(ConfigKey("[Waveform]", "DefaultZoom"),
+            WaveformWidgetRenderer::s_waveformDefaultZoom);
+    m_pWaveformZoom = std::make_unique<ControlPotmeter>(
+            ConfigKey(getGroup(), "waveform_zoom"),
+            WaveformWidgetRenderer::s_waveformMinZoom,
+            WaveformWidgetRenderer::s_waveformMaxZoom,
+            false, // allow out of bounds
+            false, // ignore no-ops
+            false, // track
+            false, // persist
+            defaultZoom);
+    const int stepCount = static_cast<int>(WaveformWidgetRenderer::s_waveformMaxZoom -
+            WaveformWidgetRenderer::s_waveformMinZoom);
+    m_pWaveformZoom->setStepCount(stepCount);
+    m_pWaveformZoom->setSmallStepCount(stepCount * 10);
+    // A wrapper for waveform_zoom_up/_down
+    m_pWaveformZoomSelector = std::make_unique<ControlEncoder>(
+            ConfigKey(getGroup(), "waveform_zoom_selector"), false);
+    connect(m_pWaveformZoomSelector.get(),
+            &ControlObject::valueChanged,
             this,
-            &BaseTrackPlayerImpl::slotWaveformZoomSetDefault);
+            [this](double value) {
+                int steps = static_cast<int>(fabs(std::round(value)));
+                const QString dir = std::round(value) > 0 ? "up" : "down";
+                while (steps > 0) {
+                    ControlObject::set(ConfigKey(getGroup(), "waveform_zoom_" + dir), 1);
+                    ControlObject::set(ConfigKey(getGroup(), "waveform_zoom_" + dir), 0);
```

Comment:
I'm not happy with this, but I couldn't make it work with ControlProxies.
I created them right after having created the ControlPotmeter and used them in the lambda, but setting the proxies would do nothing, and I have no idea why. No error or warning was thrown.